File: dc-console/src/serial/cobs.ts
// Consistent Overhead Byte Stuffing (COBS) encode/decode.
// Frames on the wire are: COBS(payload) followed by a single 0x00 delimiter.

export function cobsEncode(src: Uint8Array): Uint8Array {
  const out = new Uint8Array(src.length + Math.floor(src.length / 254) + 2);
  let readIndex = 0;
  let writeIndex = 1;
  let codeIndex = 0;
  let code = 1;

  while (readIndex < src.length) {
    if (src[readIndex] === 0) {
      out[codeIndex] = code;
      code = 1;
      codeIndex = writeIndex++;
      readIndex++;
    } else {
      out[writeIndex++] = src[readIndex++];
      code++;
      if (code === 0xff) {
        out[codeIndex] = code;
        code = 1;
        codeIndex = writeIndex++;
      }
    }
  }
  out[codeIndex] = code;
  return out.subarray(0, writeIndex);
}

// Decode a single COBS-encoded chunk (without the trailing 0x00 delimiter).
// Returns null on malformed input.
export function cobsDecode(src: Uint8Array): Uint8Array | null {
  if (src.length === 0) return null;
  const out = new Uint8Array(src.length);
  let readIndex = 0;
  let writeIndex = 0;

  while (readIndex < src.length) {
    const code = src[readIndex];
    if (code === 0 || readIndex + code > src.length) {
      return null;
    }
    readIndex++;
    for (let i = 1; i < code; i++) {
      out[writeIndex++] = src[readIndex++];
    }
    if (code !== 0xff && readIndex < src.length) {
      out[writeIndex++] = 0;
    }
  }
  return out.subarray(0, writeIndex);
}

// Streaming framer: feed bytes in any chunk size, get whole COBS-decoded
// payloads via the onFrame callback. Frames are split on 0x00 bytes.
export class FrameDecoder {
  private buf: number[] = [];

  constructor(private onFrame: (payload: Uint8Array) => void) {}

  push(chunk: Uint8Array): void {
    for (let i = 0; i < chunk.length; i++) {
      const byte = chunk[i];
      if (byte === 0x00) {
        if (this.buf.length > 0) {
          const decoded = cobsDecode(Uint8Array.from(this.buf));
          this.buf.length = 0;
          if (decoded) {
            this.onFrame(decoded);
          }
        }
      } else {
        this.buf.push(byte);
      }
    }
  }

  reset(): void {
    this.buf.length = 0;
  }
}

File: dc-firmware/src/serial/cobs.cpp
#include "cobs.hpp"

namespace cobs {

size_t encode(const uint8_t* src, size_t src_len, uint8_t* dst) {
    size_t read_index = 0;
    size_t write_index = 1;
    size_t code_index = 0;
    uint8_t code = 1;

    while (read_index < src_len) {
        if (src[read_index] == 0) {
            dst[code_index] = code;
            code = 1;
            code_index = write_index++;
            read_index++;
        } else {
            dst[write_index++] = src[read_index++];
            code++;
            if (code == 0xFF) {
                dst[code_index] = code;
                code = 1;
                code_index = write_index++;
            }
        }
    }
    dst[code_index] = code;
    return write_index;
}

size_t decode(const uint8_t* src, size_t src_len, uint8_t* dst) {
    if (src_len == 0) {
        return 0;
    }
    size_t read_index = 0;
    size_t write_index = 0;

    while (read_index < src_len) {
        uint8_t code = src[read_index];
        if (code == 0 || read_index + code > src_len) {
            return 0;  // malformed
        }
        read_index++;
        for (uint8_t i = 1; i < code; ++i) {
            dst[write_index++] = src[read_index++];
        }
        if (code != 0xFF && read_index < src_len) {
            dst[write_index++] = 0;
        }
    }
    return write_index;
}

}  // namespace cobs
